rust/src/measurements/make_private_lazyframe/group_by/mod.rs

```diff
@@ -21,136 +33,184 @@ use polars_plan::prelude::GroupbyOptions;
+                d_out = MO::add_delta(d_out, delta_joint)?;
+            } else if margin.public_info.is_none() {
+                return fallible!(FailedMap, "keys must be public if threshold is unknown");
+            }
+
+            Ok(d_out)
         }),
     )?;
 
-    let m_exprs = (t_group_by
-        >> make_basic_composition(
-            aggs.into_iter()
-                .map(|expr| {
-                    make_private_expr(
-                        expr_domain.clone(),
-                        PartitionDistance(middle_metric.clone()),
-                        output_measure.clone(),
-                        expr.clone(),
-                        global_scale,
-                    )
-                })
-                .collect::<Fallible<_>>()?,
-        )?)?;
-
-    let f_exprs = m_exprs.function.clone();
-    let privacy_map = m_exprs.privacy_map.clone();
+    t_prior >> m_group_by_agg
+}
 
-    t_prior
-        >> Measurement::new(
-            middle_domain,
-            Function::new_fallible(move |arg: &DslPlan| {
-                let mut output = plan.clone();
-                if let DslPlan::GroupBy {
-                    ref mut input,
-                    ref mut aggs,
-                    ..
-                } = output
-                {
-                    *input = Arc::new(arg.clone());
-                    *aggs = f_exprs.eval(&(arg.clone(), all()))?;
-                };
-                Ok(output)
-            }),
-            middle_metric,
-            output_measure,
-            privacy_map,
-        )?
+pub trait ApproximateMeasure: BasicCompositionMeasure {
+    fn add_delta(d_out: Self::Distance, delta_p: f64) -> Fallible<Self::Distance>;
 }
 
-#[cfg(test)]
-mod test;
+macro_rules! impl_measure_non_catastrophic {
+    ($ty:ty) => {
+        impl ApproximateMeasure for $ty {
+            fn add_delta(_d_out: Self::Distance, _delta_p: f64) -> Fallible<Self::Distance> {
+                return fallible!(
+                    MakeMeasurement,
+                    "key-sets cannot be privatized under {:?}. Approximate-DP is necessary.",
+                    stringify!($ty)
+                );
+            }
+        }
+    };
+}
+
+impl_measure_non_catastrophic!(MaxDivergence<f64>);
+impl_measure_non_catastrophic!(ZeroConcentratedDivergence<f64>);
+
+impl ApproximateMeasure for FixedSmoothedMaxDivergence<f64> {
+    fn add_delta((epsilon, delta): Self::Distance, delta_p: f64) -> Fallible<Self::Distance> {
+        Ok((epsilon, delta.inf_add(&delta_p)?))
+    }
+}
+
+fn integrate_discrete_noise_tail(
+    distribution: Distribution,
+    scale: f64,
+    tail_bound: u32,
+) -> Fallible<f64> {
+    match distribution {
+        Distribution::Laplace => integrate_discrete_laplacian_tail(scale, tail_bound),
+        Distribution::Gaussian => fallible!(
+            MakeMeasurement,
+            "gaussian tail bounds are not currently implemented"
```

Comment:
Perhaps include an issue URL here, so if someone hits this they can add their usecase where we'll see it?